File: Tests/Tests/STPCardBINMetadataTests.m
//
//  STPCardBINMetadata.m
//  StripeiOS Tests
//
//  Created by Cameron Sabol on 7/20/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPAPIClient+Private.h"
#import "STPBINRange.h"
#import "STPCardBINMetadata.h"
#import "STPTestingAPIClient.h"

@interface STPCardBINMetadataTests : XCTestCase

@end

@implementation STPCardBINMetadataTests

- (void)testAPICall {
    [[STPAPIClient sharedClient] setPublishableKey:STPTestingDefaultPublishableKey];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"Retrieve card metadata"];
    
    [[STPAPIClient sharedClient] retrieveCardBINMetadataForPrefix:@"424242" withCompletion:^(STPCardBINMetadata * _Nullable cardMetadata, NSError * _Nullable error) {
        XCTAssertNotNil(cardMetadata);
        XCTAssertTrue(cardMetadata.ranges.count > 0);
        XCTAssertNil(error);
        [expectation fulfill];
    }];
    [self waitForExpectations:@[expectation] timeout:STPTestingNetworkRequestTimeout];
}

- (void)testLoadingInBINRange {
    [[STPAPIClient sharedClient] setPublishableKey:STPTestingDefaultPublishableKey];

    XCTestExpectation *expectation = [self expectationWithDescription:@"Retrieve card metadata"];
    NSArray<STPBINRange *> *hardCodedBinRanges = [STPBINRange allRanges];
    [STPBINRange retrieveBINRangesForPrefix:@"424242" completion:^(NSArray<STPBINRange *> * _Nullable ranges, NSError * _Nullable error) {
        XCTAssertNotNil(ranges);
        XCTAssertNil(error);
        XCTAssertTrue(ranges.count > 0);
        XCTAssertTrue([STPBINRange allRanges].count == hardCodedBinRanges.count + ranges.count);
        for (STPBINRange *range in ranges) {
            XCTAssertTrue([[STPBINRange allRanges] containsObject:range]);
        }
        [expectation fulfill];
    }];
    [self waitForExpectations:@[expectation] timeout:STPTestingNetworkRequestTimeout];

}

@end

File: Tests/Tests/STPCardBINMetadataTests.swift
//
//  STPCardBINMetadata.m
//  StripeiOS Tests
//
//  Created by Cameron Sabol on 7/20/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import StripeCoreTestUtils
@testable import Stripe

class STPCardBINMetadataTests: XCTestCase {
    func testAPICall() {
        STPAPIClient.shared.publishableKey = STPTestingDefaultPublishableKey

        let expectation = self.expectation(description: "Retrieve card metadata")

        // 625035 is a randomly selected UnionPay BIN
        STPBINRange.retrieve(
            forPrefix: "625035",
            completion: { result in
                let cardMetadata = try! result.get()
                XCTAssertTrue(cardMetadata.data.count > 0)
                XCTAssertEqual(cardMetadata.data.first!.brand, .unionPay)
                expectation.fulfill()
            })
        wait(for: [expectation], timeout: STPTestingNetworkRequestTimeout)
    }

    func testLoadingInBINRange() {
        STPAPIClient.shared.publishableKey = STPTestingDefaultPublishableKey

        let expectation = self.expectation(description: "Retrieve card metadata")
        let hardCodedBinRanges = STPBINController.shared.allRanges()
        STPBINController.shared.retrieveBINRanges(forPrefix: "625035") { result in
            let ranges = try! result.get()
            XCTAssertTrue(ranges.count > 0)
            XCTAssertTrue(
                STPBINController.shared.allRanges().count == hardCodedBinRanges.count + ranges.count)
            for range in ranges {
                XCTAssertTrue(STPBINController.shared.allRanges().contains(range))
            }
            expectation.fulfill()
        }
        wait(for: [expectation], timeout: STPTestingNetworkRequestTimeout)

    }
}
